### PR TITLE
AnimeUnity fix

### PR DESCRIPTION
**Pull request by Catta1997**
Support for the new site of AnimeUnity

### CODE REVIEW

File: NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Anime.swift [hunk 1]
```diff
@@ -21,91 +21,85 @@ import Foundation
...
-                    
-                    alias: animeTitle,
+                    alias: "",
                     additionalAttributes: additionalAttributes,
-                    description: animeSynopsis,
+                    description: synopsis,
                     on: [ NASourceAnimeUnity.AnimeUnityStream: "AnimeUnity" ],
-                    episodes: [ NASourceAnimeUnity.AnimeUnityStream: episodes ],
-                    episodesAttributes: [:]
+                    episodes: [ NASourceAnimeUnity.AnimeUnityStream: episodesList.map { $0?.0 as! EpisodeLink} ]
```
Comment (SuperMarcus) [inline NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Anime.swift]: Please avoid using force casting. Use `compactMap` or similar instead if the element is optional.

File: NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Anime.swift [hunk 2]
```diff
@@ -21,91 +21,85 @@ import Foundation
...
+    struct SearchResponseAnime: Codable {
+        var to_array: [SearchResponseRecordsAnime]
+        init(from decoder: Decoder) throws {
+            var to_array = [SearchResponseRecordsAnime]()
+            var container = try decoder.unkeyedContainer()
+            while !container.isAtEnd {
+                if let route = try? container.decode(SearchResponseRecordsAnime.self) {
+                    to_array.append(route)
+                } else {
+                    _ = try? container.decode(DummyCodableAnime.self) // <-- TRICK
```
Comment (SuperMarcus) [inline NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Anime.swift:40]: Why is this necessary? Can you send me an example response that you'll have to parse with this code?

File: NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Anime.swift [hunk 3]
```diff
@@ -21,91 +21,85 @@ import Foundation
...
-                        animeTitle = String(title.dropFirst(8))
-                    }
-                    return trama
-                }
-                let animeArtworkUrl = URL(
-                    string: try bowl.select(".cover>img").attr("src")
-                    ) ?? link.image
+                let data = responseContent
+                let utf8Text = String(data: data, encoding: .utf8) ?? String(decoding: data, as: UTF8.self)
+                let  bowl = try SwiftSoup.parse(utf8Text)
```
Comment (SuperMarcus) [inline NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Anime.swift:54]:
```suggestion
                let bowl = try SwiftSoup.parse(utf8Text)
```

File: NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Anime.swift [hunk 4]
```diff
@@ -21,91 +21,85 @@ import Foundation
...
-                        let newString = episodeLink.replacingOccurrences(of: "\'", with: "%27")
-                        episodeLink = "https://animeunity.it/" + newString
-                        collection.append(.init(
-                            identifier: episodeLink,
-                            name: episodeName,
-                            server: NASourceAnimeUnity.AnimeUnityStream,
-                            parent: reconstructedAnimeLink
-                            ))
-                    }
+                let episodesList = decodedResponse.to_array.map {
```
Comment (SuperMarcus) [inline NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Anime.swift:68]: Use `compactMap` here instead -- see the comment below.

File: NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Episode.swift [hunk 5]
```diff
@@ -28,40 +27,13 @@ extension NASourceAnimeUnity {
...
-                ).tryUnwrap()
-                Log.info("[NASourceAnimeUnity] Resource found from packed scripts.")
-            } else {
-                let videourl = try videoElement.attr("src")
-                let newString = videourl.replacingOccurrences(of: "’", with: "%E2%80%99")
-                videoSource = try URL(string: newString) .tryUnwrap()
-                Log.info("[NASourceAnimeUnity] Resource found from page source.")
-            }
+            _ in
+            let video_url = link.identifier + ".mp4"
```
Comment (SuperMarcus) [inline NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Episode.swift:33]:
```suggestion
            let videoUrl = try URL(string: link.identifier + ".mp4").tryUnwrap()
```

File: NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Episode.swift [hunk 6]
```diff
@@ -28,40 +27,13 @@ extension NASourceAnimeUnity {
...
             return Episode(
                 link,
-                target: videoSource,
+                target: try video_url.asURL(),
```
Comment (SuperMarcus) [inline NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Episode.swift:36]:
```suggestion
                target: videoUrl,
```

File: NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Anime.swift [hunk 7]
```diff
@@ -21,91 +21,85 @@ import Foundation
...
-                _ = try bowl.select("div.card-body p").compactMap {entry -> String in
-                    let trama = ""
-                    if try entry.text().contains("TRAMA") {
-                        let trama = try entry.text()
-                        animeSynopsis = String(trama.dropFirst(7))
-                    }
-                    if try entry.text().contains("TITOLO") {
-                        let title = try entry.text()
-                        animeTitle = String(title.dropFirst(8))
+                _ = try bowl.select("div.info-item").compactMap {entry -> Void in
```
Comment (SuperMarcus) [inline NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Anime.swift:82]:
```suggestion
                _ = try bowl.select("div.info-item").compactMap { entry -> Void in
```

Also, consider using `reduce(into:_:)` instead of map, since you're not using the results here.

File: NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity.swift [hunk 8]
```diff
@@ -25,7 +25,85 @@ import UIKit
...
+    "&apos;": "'",
+    "&lt;": "<",
+    "&gt;": ">",
+    // HTML character entity references:
+    "&nbsp;": "\u{00a0}",
+    // ...
+    "&diams;": "♦"
+]
 
+extension String {
```
Comment (SuperMarcus) [inline NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity.swift]: Ideally, we don't want random extensions floating around (as this will confuse the other developers). So please consider the following options:
* Remove this extension and make `stringByDecodingHTMLEntitiesAnime` a member function of the `AnimeUnity` class;
* Move this extension to the file that it's being used and make it `fileprivate`;
* Move this extension to the `Utilities` group.

**Comment by SuperMarcus** [comment]
Ah, I realized that I commented on several outdated files. Please mark them as resolved and disregard if they don't apply anymore.

File: NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Anime.swift (hunk 2 above)
Comment (Catta1997) [reply NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Anime.swift:40]: I have found a better way to do it, I'll push an update

File: NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Anime.swift (hunk 7 above)
Comment (Catta1997) [reply NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Anime.swift]: I do not understand how `reduce(into:_:)` works, how I have to change it?

**Comment by Catta1997** [comment]
> @Catta1997 thanks for the PR! I've posted some comments in the code, please read through them when able. Also, I haven't seen you around in our [Discord server](https://discord.gg/dzTVzeW). Do you go by a different username there?

yes I'm in the Discord Server with the username "Edo5297"

File: NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Anime.swift (hunk 7 above)
Comment (SuperMarcus) [reply NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Anime.swift]: Check out this example: https://github.com/SuperMarcus/NineAnimator/blob/1ea990147d/NineAnimator/Models/Anime%20Source/4anime.to/FourAnime%2BFeatured.swift#L36-L51

Comment (Catta1997) [reply NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Anime.swift]:
```
var additionalAttributes = try bowl.select("div.info-item").reduce(into: [Anime.AttributeKey:Any]()) { contatiner, entry in
                    if try entry.text().contains("Anno") {
                        let year = try entry.select("small").text()
                        contatiner[.airDate] = year
                    }
                    if try entry.text().contains("Valutazione") {
                        let val = try entry.select("small").text()
                        let rating = (val as NSString).floatValue
                        contatiner[.rating] = rating
                        contatiner[.ratingScale] = Float(10.0)
                    }
                }
```
it works, thanks. I'll push a commit if it is ok

File: NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Featured.swift [hunk 9]
```diff
@@ -21,63 +21,47 @@ import Foundation
...
-                        image: artworkUrl,
-                        source: self
-                    ))
-                }
+            let data = responseContent
+            let utf8Text = String(data: data, encoding: .utf8) ?? String(decoding: data, as: UTF8.self)
+            let bowl = try SwiftSoup.parse(utf8Text)
+            var encoded = try bowl.select("the-carousel").attr("animes")
+            encoded = encoded.replacingOccurrences(of: "\n", with: "")
+            let data_json = encoded.data(using: .utf8)!
```
Comment (SuperMarcus) [inline NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Featured.swift:42]:
```suggestion
            let data_json = try encoded.data(using: .utf8).tryUnwrap()
```
Please avoid force casting whenever possible.

File: NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Anime.swift [hunk 10]
```diff
@@ -21,91 +21,75 @@ import Foundation
...
-                }
-                let animeArtworkUrl = URL(
-                    string: try bowl.select(".cover>img").attr("src")
-                    ) ?? link.image
+                let data = responseContent
+                let utf8Text = String(data: data, encoding: .utf8) ?? String(decoding: data, as: UTF8.self)
+                let bowl = try SwiftSoup.parse(utf8Text)
+                var encoded = try bowl.select("video-player").attr("anime")
+                encoded = encoded.replacingOccurrences(of: "\n", with: "")
+                let data_json = encoded.data(using: .utf8)!
```
Comment (SuperMarcus) [inline NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Anime.swift:43]:
```suggestion
                let data_json = try encoded.data(using: .utf8).tryUnwrap()
```
Please avoid force casting whenever possible.

File: NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Search.swift [hunk 11]
```diff
@@ -57,31 +62,23 @@ extension NASourceAnimeUnity {
...
-                            
-                            guard let animeUrl = URL(string: animeLinkURLString, relativeTo: endpointUrl),
-                                let coverImage = URL(string: animeArtworkPath, relativeTo: endpointUrl)
-                                else {
-                                    return nil
-                            }
-                            
+                        var encoded = try bowl.select("archivio").attr("records")
+                        encoded = encoded.replacingOccurrences(of: "\n", with: "")
+                        let data_json = encoded.data(using: .utf8)!
```
Comment (SuperMarcus) [inline NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Search.swift:67]:
```suggestion
                        let data_json = try encoded.data(using: .utf8).tryUnwrap()
```
Please avoid force casting whenever possible.

File: NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Anime.swift [hunk 12]
```diff
@@ -21,91 +21,75 @@ import Foundation
...
+                let decoder = JSONDecoder.init()
```
Comment (SuperMarcus) [inline NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Anime.swift:44]:
```suggestion
                let decoder = JSONDecoder()
```
SwiftLint or the swift compiler should give you a warning for this. Try not to use `Type.init()` for instantiation.

File: NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Episode.swift [hunk 13]
```diff
@@ -28,40 +27,13 @@ extension NASourceAnimeUnity {
             episodePageUrl in self
                 .requestManager
                 .request(url: episodePageUrl, handling: .browsing)
-                .responseString
+                .responseVoid
```
Comment (SuperMarcus) [inline NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Episode.swift:30]: Is this necessary? Seems like the results from the request are abandoned. You should directly return the `Episode` object if this network request is not necessary.

File: NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Search.swift [hunk 14]
```diff
@@ -57,31 +62,23 @@ extension NASourceAnimeUnity {
...
+                        let decoder = JSONDecoder.init()
```
Comment (SuperMarcus) [inline NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Search.swift:68]:
```suggestion
                        let decoder = JSONDecoder()
```

File: NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Featured.swift [hunk 15]
```diff
@@ -21,63 +21,47 @@ import Foundation
...
+            let decoder = JSONDecoder.init()
```
Comment (SuperMarcus) [inline NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Featured.swift:43]:
```suggestion
            let decoder = JSONDecoder()
```

File: NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Episode.swift (hunk 13 above)
Comment (Catta1997) [reply NineAnimator/Models/Anime Source/animeunity.it/AnimeUnity+Episode.swift]: Correct, removed it
